core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -484,13 +485,42 @@ class TransactionStateManager(brokerId: Int,
       responseCallback(responseError)
     }
 
-    replicaManager.appendRecords(
-      newMetadata.txnTimeoutMs.toLong,
-      TransactionLog.EnforcedRequiredAcks,
-      internalTopicsAllowed = true,
-      isFromClient = false,
-      recordsPerPartition,
-      updateCacheCallback)
+    // note that we need to grab the read lock and double check the metadata and coordinator epoch again
+    // until we have appended to log to avoid the scenario that an emigration and immigration happens after
+    // the check when handling the request
+    inReadLock(stateLock) {
+      getTransactionState(transactionalId) match {
+        case Left(err) =>
+          responseCallback(err)
+
+        case Right(Some(epochAndMetadata)) =>
+          val metadata = epochAndMetadata.transactionMetadata
+
+          metadata synchronized {
+            if (epochAndMetadata.coordinatorEpoch != coordinatorEpoch) {
+              // the coordinator epoch has changed, reply to client immediately with with NOT_COORDINATOR
+              responseCallback(Errors.NOT_COORDINATOR)
+            } else {
+              // do not need to check the metadata object itself since no current thread should be able to modify it
+              // under the same coordinator epoch, so directly append to txn log now
+
+              replicaManager.appendRecords(
```

Comment:
We actually have to be a bit careful not to introduce similar deadlocks that we had in GroupManager before. It seems a deadlock could happen in the following 2 code paths.

1. Here, we hold a lock on txnMetadata when calling replicaManager.appendRecords(), which will eventually call ProducerPurgatory.tryCompleteElseWatch() which holds a lock on the delayed produce when calling safeTryComplete(). 

2. A fetch follower request will call DelayedOperationPurgatory.checkAndComplete(), which holds a lock on the delayed produce and calls safeTryComplete(), which then holds a lock on txnMetadata.

If the caller of 1 and 2 each first holds the first lock and then proceeds, a deadlock will happen.